### PR TITLE
Move configuration rules to C library.

### PR DESCRIPTION
**Pull request by dwsteele**
Also:
- Rename most cmd/option functions and constants to what they will look like in the future when all configuration is done in the C library.
- cmd/option constants are now integers rather than strings.
- Clean up option set/load functions in unit/regression tests.

### CODE REVIEW

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 1]
```diff
@@ -0,0 +1,2234 @@
...
+        &CFGBLDOPT_RULE_TYPE => CFGBLDOPT_TYPE_BOOLEAN,
+        &CFGBLDOPT_RULE_DEFAULT => CFGBLDOPTDEF_DEFAULT_NEUTRAL_UMASK,
+        &CFGBLDOPT_RULE_COMMAND =>
+        {
+            &CFGBLDCMD_ARCHIVE_GET => {},
+            &CFGBLDCMD_ARCHIVE_PUSH => {},
+            &CFGBLDCMD_BACKUP => {},
+            &CFGBLDCMD_CHECK => {},
+            &CFGBLDCMD_EXPIRE => {},
+            &CFGBLDCMD_LOCAL => {},
```
Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Again, what happened to the CMD_INFO in this list?

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 2]
```diff
@@ -0,0 +1,2234 @@
...
+            &CFGBLDCMD_BACKUP => {},
+            &CFGBLDCMD_CHECK => {},
+            &CFGBLDCMD_INFO => {},
+            &CFGBLDCMD_LOCAL => {},
+            &CFGBLDCMD_RESTORE => {},
+            &CFGBLDCMD_STANZA_CREATE => {},
+            &CFGBLDCMD_STANZA_UPGRADE => {},
+            &CFGBLDCMD_START => {},
+            &CFGBLDCMD_STOP => {},
+            &CFGBLDCMD_EXPIRE => {},
```
Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: If we're going with "I making all these changes anyway" then can we please put cmd_expire in alphabetical order here?

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 3]
```diff
@@ -0,0 +1,2234 @@
...
+            lc(WARN)   => true,
+            lc(INFO)   => true,
+            lc(DETAIL) => true,
+            lc(DEBUG)  => true,
+            lc(TRACE)  => true
+        },
+        &CFGBLDOPT_RULE_COMMAND =>
+        {
+            &CFGBLDCMD_ARCHIVE_GET => {},
+            &CFGBLDCMD_ARCHIVE_PUS
```
Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: I see cmd_local and cmd_remote were added here - I'm assuming there was a reason for that?

Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: I can see why you added option_rule_required but I don't understand why &OPTION_RULE_DEFAULT => BACKREST_BIN,was removed.

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 4]
```diff
@@ -0,0 +1,2234 @@
...
+    push @EXPORT, qw(CFGBLDOPT_ARCHIVE_ASYNC);
+# Deprecated and to be removed
+use constant CFGBLDOPT_ARCHIVE_MAX_MB                                  => 'archive-max-mb';
+    push @EXPORT, qw(CFGBLDOPT_ARCHIVE_MAX_MB);
+use constant CFGBLDOPT_ARCHIVE_QUEUE_MAX                               => 'archive-queue-max';
+    push @EXPORT, qw(CFGBLDOPT_ARCHIVE_QUEUE_MAX);
+
+# BACKUP Section
+#-----------------------------------------------------------------------------------------------------------------------------------
+use constant CFGBLDOPT_ARCHIVE_CHECK                                => 'archive-check';
```
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: There were a few options that deviated from the general naming convention and included the command name.  This did not play well with the auto-generate of constants so was fixed.  I think it was just this, archive-copy, and restore-option that were affected.

Like I said, I'm still renaming stuff in this module and plan to reformat when I'm done.

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 5]
```diff
@@ -0,0 +1,2234 @@
...
+    push @EXPORT, qw(CFGBLDOPT_DB_INCLUDE);
+use constant CFGBLDOPT_LINK_ALL                                        => 'link-all';
+    push @EXPORT, qw(CFGBLDOPT_LINK_ALL);
+use constant CFGBLDOPT_LINK_MAP                                        => 'link-map';
+    push @EXPORT, qw(CFGBLDOPT_LINK_MAP);
+use constant CFGBLDOPT_TABLESPACE_MAP_ALL                              => 'tablespace-map-all';
+    push @EXPORT, qw(CFGBLDOPT_TABLESPACE_MAP_ALL);
+use constant CFGBLDOPT_TABLESPACE_MAP                                  => 'tablespace-map';
+    push @EXPORT, qw(CFGBLDOPT_TABLESPACE_MAP);
+use constant CFGBLDOPT_RECOVERY_OPTION                         => 'recovery-option';
```
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Same caveat.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 3 above)
Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Hmm, comment this is not showing up when I look at the review page so FYI, it is regarding CFGBLDOPT_LOG_LEVEL_STDERR.

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 6]
```diff
@@ -0,0 +1,2234 @@
...
+        }
+    },
+
+    &CFGBLDOPT_DB_TIMEOUT =>
+    {
+        &CFGBLDOPT_RULE_SECTION => CONFIG_SECTION_GLOBAL,
+        &CFGBLDOPT_RULE_TYPE => CFGBLDOPT_TYPE_FLOAT,
+        &CFGBLDOPT_RULE_DEFAULT => CFGBLDOPTDEF_DEFAULT_DB_TIMEOUT,
+        &CFGBLDOPT_RULE_ALLOW_RANGE => [CFGBLDOPTDEF_DEFAULT_DB_TIMEOUT_MIN, CFGBLDOPTDEF_DEFAULT_DB_TIMEOUT_MAX],
+        &CFGBLDOPT_RULE_COMMAND =>
```
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: CMD => false become complete obsolete some time back as far as I can see and honestly I never really understood your comment here (though it may have been valid at one time).  I change this to a hash because that's what it always gets translated to later on.

I've updated the comment to what I believe is correct.

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 7]
```diff
@@ -0,0 +1,2234 @@
...
+        {
+            &CFGBLDCMD_ARCHIVE_GET => {},
+            &CFGBLDCMD_ARCHIVE_PUSH => {},
+            &CFGBLDCMD_BACKUP => {},
+            &CFGBLDCMD_CHECK => {},
+            &CFGBLDCMD_EXPIRE => {},
+            &CFGBLDCMD_LOCAL => {},
```
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: The info command does not connect to the db so this made no sense.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 1 above)
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Ditto.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 2 above)
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Definitely.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 3 above)
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Yeah, this had to do with making the valid checking stricter.  Both commands set this option (and have for a while) so they needed it to be valid.

Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: The BACKREST_BIN constant is not available from the C library so this default must be set in configLoad() for the time being.  I plan to fix this when I move the rest of the config code into C.

File: lib/pgBackRest/Config/Config.pm [hunk 8]
```diff
@@ -18,2089 +18,103 @@ use pgBackRest::Common::Exception;
...
-    {
-        &OPTION_RULE_SECTION => CONFIG_SECTION_GLOBAL,
-        &OPTION_RULE_TYPE => OPTION_TYPE_BOOLEAN,
-        &OPTION_RULE_DEFAULT => OPTION_DEFAULT_LOG_TIMESTAMP,
-        &OPTION_RULE_COMMAND =>
-        {
-            &CMD_ARCHIVE_GET => true,
-            &CMD_ARCHIVE_PUSH => true,
-            &CMD_BACKUP => true,
-            &CMD
```
Comment (cmwshang) [inline lib/pgBackRest/Config/Config.pm]: It appears that some constants in Config.pm are "duplicated" in Data.pm. (e.g. REPO_TYPE_x == CFGBLDOPTVAL_REPO_TYPE_x, INFO_OUTPUT_x == CFGBLDOPTVAL_INFO_OUTPUT_x, BACKUP_TYPE_x == CFGBLDOPTVAL_BACKUP_TYPE_x)
Will this always be the case and if so, how do we determine what needs to be duplicated and what does not?

File: lib/pgBackRest/Config/Config.pm [hunk 9]
```diff
@@ -2216,32 +194,33 @@ sub configLoad
...
-            commandSet(CMD_HELP);
+            cfgCommandName(CFGCMD_HELP);
             return false;
         }
 
         # Validate and store options
         my $bHelp = false;
 
-        if (defined($ARGV[0]) && $ARGV[0] eq CMD_HELP && defined($ARGV[1]))
+        if (defined($ARGV[0]) && $ARGV[0] eq cfgCommandName(CFGCMD_HELP) && defined($ARGV[1]))
```
Comment (cmwshang) [inline lib/pgBackRest/Config/Config.pm]: Hmmm, when I built the docs and then entered the doc-db-master, I typed sudo -u postgres pgbackrest help check
and got:
Segmentation fault (core dumped)
but
 sudo -u postgres pgbackrest help info
 does work.

File: lib/pgBackRest/Config/Config.pm [hunk 10]
```diff
@@ -2121,9 +135,9 @@ sub configLogging
 push @EXPORT, qw(configLogging);
```
Comment (cmwshang) [inline lib/pgBackRest/Config/Config.pm]: Not sure where to enter this issue and I haven't had a chance to chase it down myself, but if I mistype the db-path as db-path1 instead of db1-path, I get the following error in addition to "Unknown option":

postgres@db-master:~$ pgbackrest --db-path1=/var/lib/postgresql/9.5/demo --stanza=demo --log-level-console=info check
Unknown option: db-path1
ERROR [124]: : process terminated due to an unhandled exception:
Use of uninitialized value $strCommand in string eq at /usr/share/perl5/pgBackRest/Config/Config.pm line 1139.
 at /usr/bin/pgbackrest line 15.
	main::__ANON__("Use of uninitialized value \$strCommand in string eq at /usr/s"...) called at /usr/share/perl5/pgBackRest/Config/Config.pm line 1139
	pgBackRest::Config::Config::cfgCommandTest(5) called at /usr/bin/pgbackrest line 40
	eval {...} called at /usr/bin/pgbackrest line 282

It works fine if I type
 pgbackrest --db1-path=/var/lib/postgresql/9.5/demo --stanza=demo --log-level-console=info check

File: build/lib/pgBackRestBuild/Config/Data.pm [hunk 11]
```diff
@@ -0,0 +1,2104 @@
...
+        }
+    },
+
+    &CFGBLDOPT_RETENTION_DIFF =>
+    {
+        &CFGBLDDEF_RULE_SECTION => CFGBLDDEF_SECTION_GLOBAL,
+        &CFGBLDDEF_RULE_TYPE => CFGBLDDEF_TYPE_INTEGER,
+        &CFGBLDDEF_RULE_REQUIRED => false,
+        &CFGBLDDEF_RULE_ALLOW_RANGE => [CFGBLDDEF_DEFAULT_RETENTION_MIN, CFGBLDDEF_DEFAULT_RETENTION_MAX],
+
```
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: These are different because they were added recently after I created the defaults (for the s3 feature).  I haven't gone back and made wholesale changes to the other options, and I definitely won't in this commit.  If we want to make changes to this logic I think that should be in another commit as well.

File: lib/pgBackRest/Config/Config.pm (hunk 8 above)
Comment (dwsteele) [inline lib/pgBackRest/Config/Config.pm]: This won't always be the case.  The duplicated constants in Config.pm will be moved to the C code and auto-generated -- that should happen before the branch is committed but I may leave it for a future commit.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 11 above)
Comment (cmwshang) [inline build/lib/pgBackRestBuild/Config/Data.pm]: It is a minimal change to remove it for this commit (I already did it to make sure it worked). I found it confusing to try to understand why the s3 stuff was not "typed". I find it much cleaner to see it displayed in the hash than buried in default logic. Stuff that's used internally (like the other defaults you add) is fine, but things that I and a user who needs to add to the config file - like is this an integer? a string? a what? I feel should be clearly and consistently laid out.

File: lib/pgBackRest/Config/Config.pm (hunk 10 above)
Comment (dwsteele) [inline lib/pgBackRest/Config/Config.pm]: Good catch.  Fixed.

File: build/lib/pgBackRestBuild/Config/Data.pm (hunk 11 above)
Comment (dwsteele) [inline build/lib/pgBackRestBuild/Config/Data.pm]: Not sure I'm convinced that this one setting should be required when everything else is option, but OK.  Fixed.

File: lib/pgBackRest/Config/Config.pm (hunk 9 above)
Comment (dwsteele) [inline lib/pgBackRest/Config/Config.pm]: Well, that's ugly.  Fixed both the segfault and the underlying issue.

**Comment by dwsteele** [comment]
Closing for now - will reopen when it is time to review again.

**Comment by dwsteele** [comment]
Go ahead and have another look now.  Please call out anything glaring, but if you have ideas for improvements let's make a list and do them as subsequent commits.  This thing is big and I'm tired of reviewing it over and over.  I'm sure you feel the same.

File: test/expect/mock-all-001.log [hunk 12]
```diff
@@ -265,7 +265,7 @@ P00  DEBUG:     Storage::Local->openWrite(): bAtomic = <false>, bPathCreate = <f
 P00  DEBUG:     Storage::Local->pathSync(): bRecurse = <false>, strPathExp = [TEST_PATH]/db-master/repo/backup/db
 P00  DEBUG:     Storage::Local->pathSync(): bRecurse = <false>, strPathExp = <REPO:BACKUP>
 P00   INFO: backup command end: completed successfully
-P00   INFO: expire command begin [BACKREST-VERSION]: --buffer-size=16384 --no-compress --config=[TEST_PATH]/db-master/pgbackrest.conf --db-timeout=45 --lock-path=[TEST_PATH]/db-master/lock --log-level-console=debug --log-level-file=trace --log-level-stderr=off --log-path=[TEST_PATH]/db-master/log --protocol-timeout=60 --repo-path=[TEST_PATH]/db-master/repo --repo-type=cifs --stanza=db
+P00   INFO: expire command begin [BACKREST-VERSION]: --buffer-size=16384 --config=[TEST_PATH]/db-master/pgbackrest.conf --lock-path=[TEST_PATH]/db-master/lock --log-level-console=debug --log-level-file=trace --log-level-stderr=off --log-path=[TEST_PATH]/db-master/log --repo-path=[TEST_PATH]/db-master/repo --repo-type=cifs --stanza=db
```
Comment (cmwshang) [inline test/expect/mock-all-001.log]: Is this right? Most times I have been seeing db- being set to db1- but here several options that were used before are not now (--no-compress, --db-timeout and --protocol-timeout)

File: lib/pgBackRest/Config/Config.pm [hunk 13]
```diff
@@ -2556,40 +448,41 @@ sub optionValidate
                 }
 
                 # If a depend value exists, make sure the option value matches
-                if ($bDependResolved && defined($$oDepend{&OPTION_RULE_DEPEND_VALUE}) &&
-                    $$oDepend{&OPTION_RULE_DEPEND_VALUE} ne $strDependValue)
+                if ($bDependResolved && cfgRuleOptionDependValueTotal($iCommandId, $iOptionId) == 1 &&
+                    cfgRuleOptionDependValue($iCommandId, $iOptionId, 0) ne $strDependValue)
```
Comment (dwsteele) [inline lib/pgBackRest/Config/Config.pm]: This is a bit weird but I want to address it when we rewrite this in C.  Right now there can be a depend value or a depend list, but a value is just a list of 1.  That's what this is hard-coded to 0.  I didn't want to create a different C function for something we'll get rid of, but also didn't want to change the Perl code too much so it would be easier to review changes.

File: lib/pgBackRest/Config/Config.pm [hunk 14]
```diff
@@ -2793,34 +682,30 @@ sub optionValidate
...
                 {
                     confess &log(ERROR, "'${strValue}' is not valid for '${strOption}' option", ERROR_OPTION_INVALID_VALUE);
                 }
 
                 # Process an allow range for the command then for the option
-                $oAllow = defined($oCommandRule) ? $$oCommandRule{&OPTION_RULE_ALLOW_RANGE} :
-                                                     $oOptionRule{$strOption}{&OPTION_RULE_ALLOW_RANGE};
-
-                if (defined($oAllow) && ($strValue < $$oAllow[0] || $strValue > $$oAllow[1]))
+                if (cfgRuleOptionAllowRange($iCommandId, $iOptionId) &&
```
Comment (dwsteele) [inline lib/pgBackRest/Config/Config.pm]: These are all instances where a command *can* override rules but doesn't currently, or the pattern of overrides allows the switch logic to optimize away command id.  For instance, cfgRuleOptionDepend uses command id but cfgRuleOptionDependOption doesn't because it has been optimized away.

File: test/expect/mock-all-001.log (hunk 12 above)
Comment (dwsteele) [inline test/expect/mock-all-001.log]: I removed them because they are not used by expire.  Mostly I was resisting the urge to tinker with options but these were so glaring.

**Comment by cmwshang** [comment]
Review complete. I did not perform an in-depth review of the Build stuff but I did not see anything glaring.

**Comment by dwsteele** [comment]
Excellent.  We'll refine as we go, but "nothing glaring" is good for now.
